src/modules/ekf2/EKF/range_finder_consistency_check.hpp

```diff
@@ -0,0 +1,77 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file range_finder_consistency_check.hpp
+ * @brief Compute statistical tests of the range finder data
+ * using the estimated velocity as a reference in order to detect sensor faults
+ */
+
+#pragma once
+
+#include <mathlib/math/filter/AlphaFilter.hpp>
+
+class RangeFinderConsistencyCheck final
+{
+public:
+	RangeFinderConsistencyCheck() = default;
+	~RangeFinderConsistencyCheck() = default;
+
+	void update(float dist_bottom, float dist_bottom_var, float vz, float vz_var, uint64_t time_us);
+
+	float getTestRatio() const { return _vel_bottom_test_ratio; }
+	float getSignedTestRatioLpf() const { return _vel_bottom_signed_test_ratio_lpf.getState(); }
+	float getInnov() const { return _vel_bottom_innov; }
+	float getInnovVar() const { return _vel_bottom_innov_var; }
+	bool isKinematicallyConsistent() const { return _is_kinematically_consistent; }
+
+private:
+	void updateConsistency(float vz, uint64_t time_us);
+
+	uint64_t _time_last_update_us{};
+	float _dist_bottom_prev{};
+
+	float _vel_bottom_test_ratio{};
+	AlphaFilter<float> _vel_bottom_signed_test_ratio_lpf{}; // average signed test ratio used to detect a bias in the data
+	float _vel_bottom_innov{};
+	float _vel_bottom_innov_var{};
+
+	bool _is_kinematically_consistent{true};
+	uint64_t _time_last_inconsistent_us{};
+
+	static constexpr float _vel_bottom_signed_test_ratio_tau = 2.f;
+	static constexpr float _vel_bottom_gate = 0.1f;
```

Comment:
This seems like a small value for the gate. If this is what is required to achieve an acceptable false positive rate, then perhaps the process or observation noise values need to be tuned instead.

Reply:
Good point, in the log I was using to tune the gate, looking at the range finder derivative and its noise (based on range finder parameter noise), we can see that the range finder noise is over-sized.
![DeepinScreenshot_select-area_20220323162211](https://user-images.githubusercontent.com/14822839/159734324-1e452387-e815-41cd-9d9a-3e906e34bc2f.png)

Tuning a bit the static noise and range-based scaler (`EKF2_RNG_NOISE` = 0.01 and `EKF2_RNG_SFE` = 0.01), we can now increase the gate to 1 to get a similar performance of the consistency check:
![DeepinScreenshot_matplotlib_20220323164320](https://user-images.githubusercontent.com/14822839/159740230-619e1757-9ec3-4863-a1f4-f3a4b214049a.png)

Reply:
@priseborough I increased the default gate size to 1 and added a note for the user in the parameter description in [ad2ae35](https://github.com/PX4/PX4-Autopilot/pull/19340/commits/ad2ae35c7b7dac8a320338420f00da22f38eb505)